runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java

```diff
@@ -758,5 +775,30 @@ public void updateLatestCheckpointIfLater(UUID checkPointId) {
             }
         }
     }
+
+    /**
+     * This queue implementation is to be used by single threaded exeuctors
+     * to restrict the amount of pending job submissions.
+     */
+    public class BoundedQueue<E> extends ArrayBlockingQueue<E> {
```

Comment:
Can't we just use the public ArrayBlockingQueue(int capacity,
                  boolean fair) directly?

Reply:
I don't believe so.

When the executor consumes the queue, it will use Queue::offer to enqueue jobs. If the queue has reached its limit, offer will return false and the ExecutorService::submit will throw a RejectedExecutionException. That will break the FastObjectLoader's logic. Ideally, we want to wait (i.e. block) until some of the pending jobs complete and free slots on the queue. By bounding the number of pending tasks, we restrict the memory usage, which is required to fix the memory "leak" issue that this PR attempts to address.